File: tests/fixtures/go/processor.go
// processor.go — data processing core with interfaces and inheritance via embedding

package fixtures

// ProcessorConfig holds pipeline configuration
type ProcessorConfig struct {
	RequiredFields []string
	Strict         bool
}

// Transformer applies a transformation to data
type Transformer interface {
	Transform(data map[string]interface{}) map[string]interface{}
}

// DataProcessor is the base pipeline processor
type DataProcessor struct {
	config       ProcessorConfig
	transformers []Transformer
}

// NewDataProcessor creates a new processor with the given config
func NewDataProcessor(cfg ProcessorConfig) *DataProcessor {
	return &DataProcessor{config: cfg}
}

// AddTransformer appends a transformer to the pipeline
func (p *DataProcessor) AddTransformer(t Transformer) {
	p.transformers = append(p.transformers, t)
}

// Process runs all transformers in sequence
func (p *DataProcessor) Process(data map[string]interface{}) map[string]interface{} {
	result := make(map[string]interface{})
	for k, v := range data {
		result[k] = v
	}
	for _, t := range p.transformers {
		result = t.Transform(result)
	}
	return result
}

// Validate checks that required fields are present
func (p *DataProcessor) Validate(data map[string]interface{}) bool {
	for _, f := range p.config.RequiredFields {
		if _, ok := data[f]; !ok {
			return false
		}
	}
	return true
}

File: tests/fixtures/python/processor.py
"""
A simple data processor module with cross-file imports.
Exercises: class extraction, method detection, import resolution, calls.
"""


class DataProcessor:
    """Processes raw data through a pipeline of transformations."""

    def __init__(self, config: dict | None = None):
        self.config = config or {}
        self._transformers: list = []

    def add_transformer(self, transformer):
        self._transformers.append(transformer)

    def process(self, data: dict) -> dict:
        result = data
        for t in self._transformers:
            result = t(result)
        return result

    def validate(self, data: dict) -> bool:
        required = self.config.get("required_fields", [])
        return all(f in data for f in required)


class JSONProcessor(DataProcessor):
    def process(self, data: dict) -> dict:
        import json
        parsed = json.loads(json.dumps(data))
        return super().process(parsed)


def create_default_processor() -> DataProcessor:
    from .filters import NormalizeFilter, RemoveNullFilter
    proc = DataProcessor({"required_fields": ["id", "name"]})
    proc.add_transformer(NormalizeFilter())
    proc.add_transformer(RemoveNullFilter())
    return proc
